config/300-subscription.yaml

```diff
@@ -38,3 +38,70 @@ spec:
     - name: Reason
       type: string
       JSONPath: ".status.conditions[?(@.type==\"Ready\")].reason"
+  validation:
+    openAPIV3Schema:
+      properties:
+        spec:
+          required:
+            - channel
+          properties:
+            channel:
+              type: object
+              required:
+                - apiVersion
+                - kind
+                - name
+              properties:
+                apiVersion:
+                  type: string
+                  minLength: 1
+                kind:
+                  type: string
+                  minLength: 1
+                name:
+                  type: string
+                  minLength: 1
+            subscriber:
+              type: object
+              properties:
```

Comment:
Is oneOf needed here?

Reply:
My reading of the [oneOf documentation](https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#oneof) is that it is used to describe that the child must conform to one of the schemas, not that one of the fields should be used. So I _think_ we could get it to work, but it would look something like:

```yaml
subscriber:
  type: object
  oneOf:
    - type: object
      required:
        - dnsName
      properties:
        dnsName:
          type: string
          minLength: 1
    - type: object
      required:
        - uri
      properties:
        uri:
          type: string
          minLength: 1
    - type: object
      required:
        - ref
      properties:
        ref:
          type: object
          required:
            - apiVersion
            - kind
            - name
          properties:
            apiVersion:
              type: string
              minLength: 1
            kind:
              type: string
              minLength: 1
            name:
              type: string
              minLength: 1
```

Reply:
You're saying this schema would allow us to automatically validate that something like:

```yaml
spec:
  subscriber:
    dnsName: a.b.c
    ref:
      apiGroup: serving.knative.dev
      kind: Service
      ref: haha
```

Is invalid? That seems nice to me.

Reply:
Having just tried this, it doesn't work. This was allowed without any complaints:

```yaml
spec:
  channel: ...
  subscriber:
    uri: foobar
    ref:
      apiVersion: v1
      kind: Service
      name: default-broker
```

I think that K8s may force the `additionalProperties: true` and change oneOf to anyOf, so it is valid for at least two of the possible objects.

Nor did it correctly block none of the fields being set.

Reply:
Thanks for testing. Sounds like `oneOf` doesn't buy us much right now.